File: A3/Sem5/PPD/LABS/lab4/ciordales/main.cpp
#include <iostream>
#include <fstream>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <cstdlib>

using namespace std;

class Barrier {
public:
    Barrier(int count) : count_(count), released_(false) {
        mutex_ = new mutex;
    }

    ~Barrier() {
        delete mutex_;
    }

    void Wait() {
        unique_lock<mutex> lock(*mutex_);
        if (--count_ == 0) {
            released_ = true;
            condition_.notify_all();
        } else {
            while (!released_) {
                condition_.wait(lock);
            }
        }
    }

private:
    mutex* mutex_;
    condition_variable condition_;
    int count_;
    bool released_;
};

void generateRandomData(int n, int m, int L) {
    string filename = "data.txt";

    ofstream writer(filename);
    writer << n << " " << m << "\n";
    srand(time(0));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; ++j) {
            writer << rand() % (L + 1) << " ";
        }
        writer << "\n";
    }
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; ++j) {
            writer << rand() % 2 << " ";
        }
        writer << "\n";
    }
    writer.close();
}

void writeMatrixToFile(const string& filename, int** V, int n, int m) {
    ofstream writer(filename);
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            writer << V[i][j] << " ";
        }
        writer << "\n";
    }
    writer.close();
}

void readMatrixFromFile(const string& filename, int n, int m, int** V) {
    ifstream reader(filename);
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            reader >> V[i][j];
        }
    }
    reader.close();
}

void runStaticLine(int n, int m, int** A, int** C) {
    auto t_start = std::chrono::high_resolution_clock::now();
    int* above = new int[m + 2];
    int* current = new int[m + 2];
    copy(A[0], A[0] + (m + 2), above);
    int left = A[1][0];

    for (int i = 1; i <= n; i++) {
        copy(A[i], A[i] + (m + 2), current);
        for (int j = 1; j <= m; j++) {
            int value = 0;

            for (int ii = 0; ii <= 1; ii++) {
                for (int jj = -1; jj <= 1; jj++) {
                    if (ii == 0 && jj == -1) {
                        value += left * C[ii + 1][jj + 1];
                    } else {
                        value += A[i + ii][j + jj] * C[ii + 1][jj + 1];
                    }
                }
            }

            for (int jj = -1; jj <= 1; jj++) {
                value += above[j + jj] * C[0][jj + 1];
            }

            if (j == m) {
                left = A[i + 1][0];
            } else {
                left = A[i][j];
            }
            A[i][j] = value;
        }
        copy(current, current + (m + 2), above);
    }

    auto t_end = std::chrono::high_resolution_clock::now();

    double elapsed_time_ms = std::chrono::duration<double, std::milli>(t_end - t_start).count();

    cout << elapsed_time_ms;

    writeMatrixToFile("outSecvential.txt", A, n + 2, m + 2);

    delete[] above;
    delete[] current;
}

void runThreadsLine(int n, int m, int** A, int** C, int p, Barrier& barrier) {
    thread* threads = new thread[p];
    int rest = n % p;
    int cat = n / p;
    int start = 1;
    int end;

    auto t_start = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < p; i++) {
        end = start + cat;
        if (rest > 0) {
            end++;
            rest--;
        }
        threads[i] = thread([n, m, i, start, end, &C, &A, &barrier]() {
            int* above = new int[m + 2];
            int* current = new int[m + 2];
            int* frontier_above = new int[m + 2];
            int* frontier_below = new int[m + 2];

            copy(A[start - 1], A[start - 1] + (m + 2), frontier_above);
            copy(A[end], A[end] + (m + 2), frontier_below);

            barrier.Wait();

            copy(A[start], A[start] + (m + 2), above);
            int left = A[start][0];

            //first line
            for (int j = 1; j <= m; j++) {
                int value = 0;

                for (int ii = 0; ii <= 1; ii++) {
                    for (int jj = -1; jj <= 1; jj++) {
                        if (ii == 0 && jj == -1) {
                            value += left * C[ii + 1][jj + 1];
                        } else {
                            value += A[start + ii][j + jj] * C[ii + 1][jj + 1];
                        }
                    }
                }

                for (int jj = -1; jj <= 1; jj++) {
                    value += frontier_above[j + jj] * C[0][jj + 1];
                }

                left = A[start][j];
                A[start][j] = value;
            }

            left = A[start + 1][0];

            //inner lines
            for (int i = start + 1; i < end - 1; i++) {
                copy(A[i], A[i] + (m + 2), current);

                for (int j = 1; j <= m; j++) {
                    int value = 0;

                    for (int ii = 0; ii <= 1; ii++) {
                        for (int jj = -1; jj <= 1; jj++) {
                            if (ii == 0 && jj == -1) {
                                value += left * C[ii + 1][jj + 1];
                            } else {
                                value += A[i + ii][j + jj] * C[ii + 1][jj + 1];
                            }
                        }
                    }

                    for (int jj = -1; jj <= 1; jj++) {
                        value += above[j + jj] * C[0][jj + 1];
                    }

                    if (j == m) {
                        left = A[i + 1][0];
                    } else {
                        left = A[i][j];
                    }
                    A[i][j] = value;
                }

                copy(current, current + (m + 2), above);
            }

            //last line
            for (int j = 1; j <= m; j++) {
                int value = 0;

                for (int jj = -1; jj <= 1; jj++) {
                    value += above[j + jj] * C[0][jj + 1];
                }

                for (int jj = -1; jj <= 1; jj++) {
                    if (jj == -1) {
                        value += left * C[1][jj + 1];
                    } else {
                        value += A[end - 1][j + jj] * C[1][jj + 1];
                    }
                }

                for (int jj = -1; jj <= 1; jj++) {
                    value += frontier_below[j + jj] * C[2][jj + 1];
                }

                left = A[end - 1][j];
                A[end - 1][j] = value;
            }

            delete[] above;
            delete[] current;
            delete[] frontier_above;
            delete[] frontier_below;
        });
        start = end;
    }

    for (int i = 0; i < p; i++) {
        threads[i].join();
    }

    auto t_end = std::chrono::high_resolution_clock::now();

    double elapsed_time_ms = std::chrono::duration<double, std::milli>(t_end - t_start).count();
    cout << elapsed_time_ms << endl;

    writeMatrixToFile("outThreadsLine.txt", A, n + 2, m + 2);

    delete[] threads;
}

void borderMatrix(int** A, int n, int m) {
    // Fill the top border
    copy(A[1], A[1] + (m + 2), A[0]);

    // Fill the right border
    for (int i = 0; i < n + 2; i++) {
        A[i][m + 1] = A[i][m];
    }

    // Fill the left border
    for (int i = 0; i < n + 2; i++) {
        A[i][0] = A[i][1];
    }

    // Fill the bottom border
    copy(A[n], A[n] + (m + 2), A[n + 1]);
}

void solveThreadsLine(int n, int m, int p, int **A, int **C) {
    Barrier barrier(p);
    runThreadsLine(n, m, A, C, p, barrier);

    int** AA;
    AA = new int* [n + 2];
    for (int i = 0; i < n + 2; i++) {
        AA[i] = new int[m + 2];
    }
    readMatrixFromFile("outSecvential.txt", n + 2, m + 2, AA);

    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= m; ++j) {
            if (A[i][j] != AA[i][j])
                cerr << "A is not equal to AA at position: (" << i << "," << j << ")\n";
        }
    }
}

int main() {
//    generateRandomData(10000, 10000, 5000);
    int n, m, p = 16;
    int** A, ** C;
    string fileName = "data.txt";
    ifstream reader(fileName);
    reader >> n >> m;
    A = new int* [n + 2];
    for (int i = 0; i < n + 2; i++) {
        A[i] = new int[m + 2];
    }
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= m; j++) {
            reader >> A[i][j];
        }
    }
    C = new int* [3];
    for (int i = 0; i < 3; i++) {
        C[i] = new int[3];
        for (int j = 0; j < 3; j++) {
            reader >> C[i][j];
        }
    }
    reader.close();
    borderMatrix(A, n, m);

//    runStaticLine(n, m, A, C);

    solveThreadsLine(n, m, p, A, C);

    for (int i = 0; i < n + 2; i++) {
        delete[] A[i];
    }
    delete[] A;

    for (int i = 0; i < 3; i++) {
        delete[] C[i];
    }
    delete[] C;

    return 0;
}

File: A3/Sem5/PPD/LABS/lab4/ciordales/Main.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.Random;
import java.io.FileWriter;
import java.io.IOException;
import java.util.concurrent.BrokenBarrierException;
import java.util.concurrent.CyclicBarrier;

public class Main {

    public static void main(String[] args) throws IOException {
//        generateRandomData(10000, 10000, 5000);

        int n, m, p = 16;
        int[][] A, C;
        //p = Integer.parseInt(args[0]); // if i want the number of threads to be read from script

        String fileName = "C:\\Users\\Demian\\Desktop\\Programare paralela\\Laborator1_PPD\\Laborator1_PPD_Java\\data.txt";
        BufferedReader br = new BufferedReader(new FileReader(fileName));
        String[] dimensions = br.readLine().split(" ");
        n = Integer.parseInt(dimensions[0]);
        m = Integer.parseInt(dimensions[1]);
        A = new int[n + 2][m + 2];
        for (int i = 1; i <= n; i++) {
            String[] elements = br.readLine().split(" ");
            for (int j = 1; j <= m; j++)
                A[i][j] = Integer.parseInt(elements[j - 1]);//read the inner matrix
        }
        C = new int[3][3];
        for (int i = 0; i < 3; i++) {
            String[] elements = br.readLine().split(" ");
            for (int j = 0; j < 3; j++)
                C[i][j] = Integer.parseInt(elements[j]);
        }

        borderMatrix(A);

//        runStaticLine(n, m, A, C);

        solveThreadsLine(n, m, A, C, p);
    }

    public static class MyThreadLine extends Thread {
        int id, start, end, n, m;
        int[][] C, A, V;
        CyclicBarrier barrier;

        public MyThreadLine(int n, int m, int id, int start, int end, int[][] C, int[][] A, CyclicBarrier barrier) {
            this.id = id;
            this.end = end;
            this.start = start;
            this.C = C;
            this.A = A;
            this.n = n;
            this.m = m;
            this.barrier = barrier;
        }

        @Override
        public void run() {
            int[] frontier_above = new int[m + 2];
            int[] frontier_below = new int[m + 2];
            System.arraycopy(A[start - 1], 0, frontier_above, 0, m + 2);
            System.arraycopy(A[end], 0, frontier_below, 0, m + 2);

            try {
                barrier.await();
            } catch (InterruptedException | BrokenBarrierException ignored) {
            }

            int[] above = new int[m + 2];
            int[] current = new int[m + 2];
            System.arraycopy(A[start], 0, above, 0, m + 2);
            int left = A[start][0];

            //first line
            for (int j = 1; j <= m; j++) {
                int value = 0;

                for (int ii = 0; ii <= 1; ii++)
                    for (int jj = -1; jj <= 1; jj++)
                        if (ii == 0 && jj == -1)
                            value += left * C[ii + 1][jj + 1];
                        else
                            value += A[start + ii][j + jj] * C[ii + 1][jj + 1];

                for (int jj = -1; jj <= 1; jj++)
                    value += frontier_above[j + jj] * C[0][jj + 1];

                left = A[start][j];
                A[start][j] = value;
            }

            left = A[start + 1][0];
            //inner lines
            for (int i = start + 1; i < end - 1; i++) {
                System.arraycopy(A[i], 0, current, 0, m + 2);
                for (int j = 1; j <= m; j++) {
                    int value = 0;

                    for (int ii = 0; ii <= 1; ii++)
                        for (int jj = -1; jj <= 1; jj++)
                            if (ii == 0 && jj == -1)
                                value += left * C[ii + 1][jj + 1];
                            else
                                value += A[i + ii][j + jj] * C[ii + 1][jj + 1];

                    for (int jj = -1; jj <= 1; jj++)
                        value += above[j + jj] * C[0][jj + 1];

                    if (j == m)
                        left = A[i + 1][0];
                    else
                        left = A[i][j];
                    A[i][j] = value;
                }
                System.arraycopy(current, 0, above, 0, m + 2);
            }

            //last line
            for (int j = 1; j <= m; j++) {
                int value = 0;

                for (int jj = -1; jj <= 1; jj++)
                    value += above[j + jj] * C[0][jj + 1];

                for (int jj = -1; jj <= 1; jj++)
                    if (jj == -1)
                        value += left * C[1][jj + 1];
                    else
                        value += A[end - 1][j + jj] * C[1][jj + 1];

                for (int jj = -1; jj <= 1; jj++)
                    value += frontier_below[j + jj] * C[2][jj + 1];

                left = A[end - 1][j];
                A[end - 1][j] = value;
            }
        }
    }

    public static void runStaticLine(int n, int m, int[][] A, int[][] C) {
        long startTime = System.nanoTime();
        int[] above = new int[m + 2];
        int[] current = new int[m + 2];
        System.arraycopy(A[0], 0, above, 0, m + 2);
        int left = A[1][0];

        for (int i = 1; i <= n; i++) {
            System.arraycopy(A[i], 0, current, 0, m + 2);
            for (int j = 1; j <= m; j++) {
                int value = 0;

                for (int ii = 0; ii <= 1; ii++)
                    for (int jj = -1; jj <= 1; jj++)
                        if (ii == 0 && jj == -1)
                            value += left * C[ii + 1][jj + 1];
                        else
                            value += A[i + ii][j + jj] * C[ii + 1][jj + 1];

                for (int jj = -1; jj <= 1; jj++)
                    value += above[j + jj] * C[0][jj + 1];

                if (j == m)
                    left = A[i + 1][0];
                else
                    left = A[i][j];
                A[i][j] = value;
            }
            System.arraycopy(current, 0, above, 0, m + 2);
        }

        System.out.println((System.nanoTime() - startTime) / 1E6);

        writeMatrixToFile("outSecvential.txt", A);
    }

    private static void solveThreadsLine(int n, int m, int[][] A, int[][] C, int p) {
        runThreadsLine(n, m, A, C, p);

        int[][] AA;
        AA = new int[n + 2][m + 2];
        readMatrixFromFile("outSecvential.txt", n + 2, m + 2, AA);
        boolean areEqual = true;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                if (A[i][j] != AA[i][j]) {
                    areEqual = false;
                    break;
                }
            }
        }
        if (!areEqual) {
            throw new RuntimeException("A is not equal to AA");
        }
    }

    public static void runThreadsLine(int n, int m, int[][] A, int[][] C, int p) {
        Thread[] threads = new Thread[p];
        int rest = n % p;
        int cat = n / p;
        int start = 1;
        int end;

        long startTime = System.nanoTime();

        CyclicBarrier barrier = new CyclicBarrier(p);
        for (int i = 0; i < p; i++) {
            end = start + cat;
            if (rest > 0) {
                end++;
                rest--;
            }
            threads[i] = new MyThreadLine(n, m, i, start, end, C, A, barrier);
            threads[i].start();
            start = end;
        }

        for (int i = 0; i < p; i++) {
            try {
                threads[i].join();
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }

        System.out.println((System.nanoTime() - startTime) / 1E6);

        writeMatrixToFile("outThreadsLine.txt", A);
    }

    public static void borderMatrix(int[][] A) {
        int fullN = A.length;
        int fullM = A[0].length;

        // Fill the top border
        System.arraycopy(A[1], 0, A[0], 0, fullM);

        // Fill the right border
        for (int i = 0; i < fullN; i++)
            A[i][fullM - 1] = A[i][fullM - 2];

        // Fill the left border
        for (int i = 0; i < fullN; i++)
            A[i][0] = A[i][1];

        // Fill the bottom border
        System.arraycopy(A[fullN - 2], 0, A[fullN - 1], 0, fullM);

    }

    public static void generateRandomData(int n, int m, int L) {
        String filename = "data" + ".txt";

        int[][] A = new int[n][m];
        int[][] conv = new int[3][3];

        Random rand = new Random();
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                A[i][j] = rand.nextInt(L);

        for (int i = 0; i < 3; i++)
            for (int j = 0; j < 3; j++)
                conv[i][j] = rand.nextInt(2);

        try (FileWriter writer = new FileWriter(filename)) {
            writer.write(n + " " + m + "\n");
            for (int[] row : A) {
                for (int element : row)
                    writer.write(element + " ");
                writer.write("\n");
            }
            for (int[] row : conv) {
                for (int element : row)
                    writer.write(element + " ");
                writer.write("\n");
            }
        } catch (IOException e) {
            System.err.println("Error writing to file: " + e.getMessage());
        }
    }

    public static void writeMatrixToFile(String filename, int[][] V) {
        try (FileWriter writer = new FileWriter(filename)) {
            for (int[] row : V) {
                for (int element : row)
                    writer.write(element + " ");
                writer.write("\n");
            }
        } catch (IOException e) {
            System.err.println("Error writing to file: " + e.getMessage());
        }
    }

    public static void readMatrixFromFile(String filename, int n, int m, int[][] V) {
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            for (int i = 0; i < n; i++) {
                String[] elements = br.readLine().split(" ");
                for (int j = 0; j < m; j++)
                    V[i][j] = Integer.parseInt(elements[j]);
            }
        } catch (IOException e) {
            System.err.println("Error reading from file: " + e.getMessage());
        }
    }
}
